### PR TITLE
[WIP] [3.3.5] Core/Battlefield: Wintergrasp rewrite

### PR DESCRIPTION
**Pull request by ccrs**
What the tittle says.
tagging many issues

**Target branch(es):** 3.3.5

### CODE REVIEW

**Comment by ccrs** [comment]
@tkrokli help me find all the related issues please

**Comment by tkrokli** [comment]
Making a list here if you want to import it into your PR comment post:
- [ ] **[wintergrasp] Game Over - prematurely #6132**
- [ ] **[wintergrasp] No announcements on wall breach #6133**
- [ ] **Wintergrasp honor reward is wrong #8413**
- [ ] **BattleField/Wintergrasp: Bug Workshop #8474**
- [ ] **[Battleground] Wintergrasp walls, doors and titan relic disappear after several runs #8661**
- [ ] **Wintergrasp Spirit Guide not working #9035**
- [ ] **Wintergrasp Walls & Workshops - Visibility Range #9205**
- [ ] **Wintergrasp honor and mark of honor bug #9423**
- [ ] **[3.3.5] Character flash at WG (Winter Tower Edge) #19241**
- [ ] **

(if you want to include the tick boxes, just edit, copy and cancel)

Other hits besides these listed above are only indirectly related to BF/WG, like Calendar issues.

**Comment by ccrs** [comment]
thanks!

**Comment by Adizbek** [comment]
It would great if you implement WB.👍👍👍

File: src/server/game/Battlefield/Battlefield.h [hunk 1]
```diff
@@ -101,321 +206,87 @@ class TC_GAME_API BfCapturePoint
...
-
-    protected:
-        TeamId m_ControlTeam;
-        uint32 m_GraveyardId;
-        ObjectGuid m_SpiritGuide[BG_TEAMS_COUNT];
-        GuidSet m_ResurrectQueue;
-        Battlefield* m_Bf;
-};
-
-class TC_GAME_API Battlefield : public ZoneScript
```
Comment (joschiwald) [inline src/server/game/Battlefield/Battlefield.h]: you don't like documentation?

File: src/server/game/Battlefield/BattlefieldMgr.h [hunk 2]
```diff
@@ -19,46 +19,36 @@
...
         static BattlefieldMgr* instance();
 
-        // create battlefield events
-        void InitBattlefield();
+        // create battlefields
+        void Initialize();
 
         // called when a player enters an battlefield area
-        void HandlePlayerEnterZone(Player* player, uint32 zoneId);
+        void HandlePlayerEnterZone(Player* player, uint32 mapId, uint32 zoneId);
```
Comment (joschiwald) [inline src/server/game/Battlefield/BattlefieldMgr.h]: zoneid is unique, why do you need a mapid here?

File: src/server/game/Battlefield/Battlefield.h (hunk 1 above)
Comment (ccrs) [inline src/server/game/Battlefield/Battlefield.h]: u what? I've been updating every old and wrong comment and every log
the moment I start updating battlefield class, is when I'll put the doc

File: src/server/game/Battlefield/BattlefieldMgr.h (hunk 2 above)
Comment (ccrs) [inline src/server/game/Battlefield/BattlefieldMgr.h]: unique for different maps?

File: src/server/game/Battlefield/Battlefield.h (hunk 1 above)
Comment (joschiwald) [inline src/server/game/Battlefield/Battlefield.h]: you almost deleted all comments from this class

**Comment by ccrs** [comment]
I agree, but there are many points here to take into account.

1. The old code is simply a huge pile of crap, copy pasted from outdoorpvp, which is also not the finest implementation, and some random corrections using bg logic
2. My code is pretty much done and already implemented, i wont go through the process again, first because i dont have the time and second because i dont want to
3. The number of changes that are necessary would make this impossible to review even following the steps on a single PR, and we go back to point 2

At least im trying to make it simple and as easy as possible to review, separating it into commits and so on.

If a merge to master is to fear here, i cant do much about it.

**Comment by Ascathor** [comment]
I like the concept of you rehauling and refactoring the battleground system.

**Comment by ccrs** [comment]
battlefield*

Dunnow if its a joke or a compliment though

**Comment by Warpten** [comment]
Battlegrounds are beyond saving, it sadly is not an easy thing to rewrite properly due to the sheer amount of hacks used to tie them to map events (such as flag capture, for starters - I won't even mention worldstates). But this is outside the scope of this PR

**Comment by ccrs** [comment]
positions taken from the original code... sniffed ones are welcome

**Comment by ccrs** [comment]
image

improving boyz

File: src/server/worldserver/worldserver.conf.dist [hunk 3]
```diff
@@ -3816,9 +3816,9 @@ Logger.mmaps=3,Server
 #Logger.ahbot=3,Console Server
 #Logger.auctionHouse=3,Console Server
 #Logger.bg.arena=3,Console Server
-#Logger.bg.battlefield=3,Console Server
 #Logger.bg.battleground=3,Console Server
 #Logger.bg.reportpvpafk=3,Console Server
+#Logger.battlefield=3,Console Server
```
Comment (Raydor) [inline src/server/worldserver/worldserver.conf.dist]: This is mindly infuriating, please order it alphabetically before my mind explodes >.<

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 4]
```diff
@@ -203,48 +205,132 @@ class npc_wg_spirit_guide : public CreatureScript
...
                                 player->TeleportTo(safeLoc->map_id, safeLoc->x, safeLoc->y, safeLoc->z, 0);
                 }
+
                 return true;
             }
+
+            void UpdateAI(uint32 diff) override
+            {
+                if (!me->HasUnitState(UNIT_STATE_CASTING))
+                    DoCast(me, SPELL_CHANNEL_SPIRIT_HEAL);
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: why not DoCastSelf()?

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 5]
```diff
@@ -338,6 +417,59 @@ class npc_wg_queue : public CreatureScript
...
+                                else
+                                {
+                                    if (!me->FindNearestGameObject(GO_WINTERGRASP_PORTAL, 25.f))
+                                        me->SummonGameObject(GO_WINTERGRASP_PORTAL, PortalPos[1], QuaternionData(), 1000);
+                                    _events.Repeat(Minutes(5));
+                                }
+                            }
+                            break;
+                        case EVENT_FROST_ARMOR:
+                            DoCast(me, SPELL_FROST_ARMOR);
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: DoCastSelf()

File: src/server/scripts/Northrend/zone_wintergrasp.cpp [hunk 6]
```diff
@@ -338,6 +417,59 @@ class npc_wg_queue : public CreatureScript
...
+                            _events.Repeat(180000);
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: Missing chrono here :)

Comment (ccrs) [inline src/server/scripts/Northrend/zone_wintergrasp.cpp]: I'm an oldfashion hipster, ty

**Comment by ccrs** [comment]
oook, it builds
time to test boyz

File: src/server/game/Miscellaneous/Language.h [hunk 7]
```diff
@@ -1228,42 +1237,42 @@ enum TrinityStrings
...
-    LANG_CREATURE_NO_INTERIOR_POINT_FOUND         = 11011,
-    LANG_CREATURE_MOVEMENT_NOT_BOUNDED            = 11012,
-    LANG_CREATURE_MOVEMENT_MAYBE_UNBOUNDED        = 11013,
-    LANG_INSTANCE_BIND_MISMATCH                   = 11014,
-    LANG_CREATURE_NOT_AI_ENABLED                  = 11015,
-    LANG_SELECT_PLAYER_OR_PET                     = 11016,
-    LANG_SHUTDOWN_DELAYED                         = 11017,
-    LANG_SHUTDOWN_CANCELLED                       = 11018
+    LANG_BAN_ACCOUNT_YOUBANNEDMESSAGE_WORLD = 11006,
+    LANG_BAN_ACCOUNT_YOUPERMBANNEDMESSAGE_WORLD = 11007,
```
Comment (tkrokli) [inline src/server/game/Miscellaneous/Language.h]: Is there any improvement to be gained from removing the enum alignment?
To me, it looks less readable than before, but maybe I have missed something.

**Comment by Keader** [comment]
https://github.com/TrinityCore/TrinityCore/pull/20144.diff

$ git apply wg.diff
wg.diff:98: trailing whitespace.
-- Horde Workshop Guards
warning: 1 line adds whitespace errors.

File: src/server/game/Miscellaneous/Language.h (hunk 7 above)
Comment (Keader) [inline src/server/game/Miscellaneous/Language.h]: he should change it before merge xD

Comment (ccrs) [inline src/server/game/Miscellaneous/Language.h]: Consistency with the rest, if it has change, change it all together, but I wont bother with that

**Comment by ccrs** [comment]
Apply it, the diff, on a clean TC please; will correct the whitespace later

**Comment by Keader** [comment]
1 - Why Tower Cannon is all dead? (in fortress)
<details>

image

</details>

2 - I found a exploit with  58729 [Spiritual Immunity] .
If you die and spirit ress you, you get Spiritual Immunity aura.
Correct behavior: you lost Spiritual Immunity if you not near of Spirit
Exploit: If you use .tele or get summon, you not lost Spiritual Immunity aura.

3 - I cant break the Fortress main door. Spells like Plague Barrel cant hit the object (i cant damage him)

4 - Plagued Barrel used by Catapult, the trajectory of barrel animation is super wrong (is just cosmetic stuff)

I can test more tomorrow =)

**Comment by Ascathor** [comment]
@ccrs My bad for the battleground/battlefield messup. I was honest though, I like it.

**Comment by nyhmii** [comment]
How do i apply the diff with the whitespace ? manually???

**Comment by ccrs** [comment]
its just one lane, correct it or wait till I commit the correction
or even better dont use a diff and get https://github.com/ccrs/TrinityCore/tree/wintergrasp

**Comment by ccrs** [comment]
@Keader 
1, 3 - will take a look
2, 4 - this type of things, achievements, spells, etc, are a kinda out of the scope of the PR
5 - can you confirm from retail that bases should all be on defender control?

**Comment by Keader** [comment]
> 5 - can you confirm from retail that bases should all be under defender control?

It's correct, tested today in retail.
When battle start, all defense bases is in your control.

I noticed in retail, that when you make some vehicle (in defense bases), you automatic mount on him, it not happen on trinity.

I will test current state of this PR tomorrow xD

**Comment by nyhmii** [comment]
Is this a release "so far" that is working to test and apply to our main 3.3.5 core?

**Comment by lachtanek** [comment]
Looks cool, a few things we noticed:
1. Bottom towers, graveyards and workshops should be owned by attacking fraction
2. When 2nd WG started, the object in the main room which you click to win didn't appear
We couldn't really test towers since they're weird, they look like all deffender-controlled on minimap, but some of them could be destroyed by deffender and some by attacker.

**Comment by nyhmii** [comment]
Wintergrasp RESETS after 2nd battle. and then it's a INSTANT win. I checked the COnfig and settings are correct. Keeps active for Horde. EVERYTHING is active for horde. Once WG starts, Horde still wins within start time.

**Comment by ccrs** [comment]
@lachtanek thanks for the feedback!
I'll take a look into the second issue reported as soon as I have some spare time
The first one seems invalid though, @Keader checked and confirmed that on retail every base is controlled by the defeneder on start.

@nyhmii I'll look at it, though in my tests I couldnt reproduce it

**Comment by lachtanek** [comment]
@ccrs From http://wowwiki.wikia.com/wiki/Wintergrasp

> Towers
The keep has four towers as part of its design, at the corners of walls. The attackers have three towers at the southern end of the map. For each of the attacking team's towers that the defenders manage to destroy, the defenders will be rewarded with bonus honor.
...
As of patch 3.0.8, each of the attacking team's towers provide a Tower Control buff, which is a 5% bonus to damage
...
If all three towers are destroyed, the duration of the battle is shortened by 10 minutes. As the game starts with a 30 minute countdown, this could mean the difference between defeat or victory.

and

> Defender basic strategy
As defenders you need to divide your forces. ... If you can keep the attackers from capturing and holding the nearby vehicle factories, you gain a huge strategic advantage. They either need to spend a lot of forces capturing them, or drive vehicles up from the south which is slow.

About the issue of WG not working the second time, couldn't reproduce it either.
If you need help with any coding, hit me up.

**Comment by Keader** [comment]
Before WG start:
<details>

wowscrnshot_082117_172601

</details>

After WG start:
<details>

wowscrnshot_082117_174724

</details>

**Comment by ccrs** [comment]
erm Keader, thats the opposite of what you told me xD
I'll rever the commit I guess

**Comment by ccrs** [comment]
@Keader 
> 5 - When battle start, with Horde control, ally should take other bases right?
With your PR, other bases is with ally control when WG start

that evidence you just posted shows the opposite xd

@lachtanek care about those notes, WG had several major changes on retail
and sure, hit IRC anytime, I'm usually connected

**Comment by Keader** [comment]
lul i told you here: https://github.com/TrinityCore/TrinityCore/pull/20144#issuecomment-323867279

> When battle start, all defense bases is in your control. (defense team)

**Comment by ccrs** [comment]
then what do those images mean? you are showing a wintergrasp with all bases taken by the attacker team

**Comment by ccrs** [comment]
that targets weird damage behaviour for buildings

**Comment by ccrs** [comment]
oook thats everything
last thing is the queue, the current implementation is completely wrong, cannot even call it a "queue"
when I'm done with that and if you guys cant find anything else, I'll consider it ready to merge

leaving spells achievements and creature scripts away from the PR, this is big enough already

**Comment by nyhmii** [comment]
After applying latest fixes:

wowscrnshot_082517_203925

Still owned by horde completely:

wowscrnshot_082517_203933

**Comment by Keader** [comment]
> then what do those images mean? you are showing a wintergrasp with all bases taken by the attacker team

In 2ª Screenshot, i was alone in wintergrasp.
And when battle started, all bases was in my control (attack team)
Only fortress in control of defense team.

**Comment by nyhmii** [comment]
@Keader  Bottom 3 Towers should belong to Alliance? when battle starts? in this example. Because when WG starts, Alliance has to break down towers or defend their towers?

or am i getting it all wrong here?

Another thing.

Opposite team(alliance troops/vendors spawn in the MAIN base when battle starts. which is wrong.

**Comment by Keader** [comment]
@nyhmii Alliance has control of WG when battle start right?
So Horde need defend this 3 bottom towers, and alliance should destroy.
If alliance destroy this 3 towers, battle time (time to horde take relic) is reduced.

**Comment by nyhmii** [comment]
@Keader  So in terms of what you said, When the battle starts, will alliance get the bottom 3 towers or horde? because currently, Horde is in control, when the battle starts horde has control of ALL the towers and bases.

**Comment by Keader** [comment]
Horde (defense team)

**Comment by nyhmii** [comment]
@Keader 
@ccrs 
@tkrokli 

Are you willing to make a cross Faction BG script for this rev(have n old one) in return for some bounty?

I'm struggling to carry it over to this rev as i am more java/c# 

Jean.idohomes@gmail.com if interested.

**Comment by nyhmii** [comment]
After battle : Horde Controlling. 3 bottom towers display not fixed. and Horde destroyed their own towers???


image

**Comment by ccrs** [comment]
The state of the gameobjects must be kept till next battle, no issue there.
* If its not war time:
all workshops and towers belong to defender team (if destroyed, stay destroyed).
* On battle start:
workshops and outside towers team is changed to attacker and the linked gameobject is rebuilt.

Thats retail behaviour and thats the current behaviour of the rework (latest commits)
So end of the discussion if no clear evidence is provided.

**Comment by nyhmii** [comment]
@ccrs tested this, it seams legit only problem. if no one is in WG que. the timer auto resets... so 30mins don't go past.

to reproduce the problem:
1. log in dev/gm mode (invis)
2. No one in WG, .tele wintergrasp.
3. wait a few sec's and WG won by previous winner.

**Comment by nyhmii** [comment]
@ccrs i saw https://github.com/TrinityCore/TrinityCore/issues/13118#issuecomment-115344762

I'm willing to pay you a decent bounty. for latest rev.

**Comment by nyhmii** [comment]
PROBLEM 10 people tested 👍 

Horde is in Control. but alliance can't damage any walls.

image


image

**Comment by nyhmii** [comment]
READ COMMENTS 👍 
image
SOUTHERN TOWERS OWNED BY ALLIANCE 👍 
Look at this:
image

**Comment by nyhmii** [comment]
LATEST ONE 👍 
Players cannot damage any walls. Since it's SHOWN as their OWN walls
Alliance can't attack alliance walls.

image

**Comment by lachtanek** [comment]
Did you try that with latest rev btw?

**Comment by ccrs** [comment]
@nyhmii that doesnt make any sense at all, I've never seen that happenin. How did u apply the PR to your core? Are you using the latest commit?
@lachtanek can you reproduce what he is reporting?

**Comment by lachtanek** [comment]
@ccrs We tried like 2-3 times with 2-3 ppl but it seemed to work as expected, so no. I may try more tests when I have time tomorrow.

**Comment by nyhmii** [comment]
Latest test : 
Horde can't attack their own walls to win

image

**Comment by nyhmii** [comment]
Currently Alliance Controlled in battle -
(it's so mixed up) aiai

image

**Comment by nyhmii** [comment]
@ccrs I downloaded the and cloned and Git Synced

**Comment by nyhmii** [comment]
@ccrs  went through the code manually.... same as yours

**Comment by nyhmii** [comment]
Redownloaded the Fresh core, and commits... still the same.

It's stable for like the 1st 10 times and then it starts to bug out

To test WG : timer to 1 min

**Comment by ccrs** [comment]
I cant reproduce in any way what @nyhmii is reporting (actually feels like he is using just a clean core) even after 2 hours of intensive testing (around 100 battles) I'm marking it as invalid.

**Comment by nyhmii** [comment]
@ccrs I downloaded the Source directly from your tree... Maybe crossfaction groups can mess this up with that setting on?

File: src/server/game/Battlefield/Battlefield.cpp [hunk 8]
```diff
@@ -267,26 +267,19 @@ void Battlefield::InvitePlayersInQueueToWar()
 {
     for (uint8 team = 0; team < PVP_TEAMS_COUNT; ++team)
     {
-        for (auto itr = _playerQueue[team].begin(); itr != _playerQueue[team].end();)
+        for (auto itr = _playerQueue[team].begin(); itr != _playerQueue[team].end(); itr = _playerQueue[team].erase(itr))
```
Comment (Shauren) [inline src/server/game/Battlefield/Battlefield.cpp]: do you *ALWAYS* erase? then don't erase at all, .clear()

Comment (ccrs) [reply src/server/game/Battlefield/Battlefield.cpp]: erase till the container is empty or cant invite cause battlefield is full

Comment (ccrs) [reply src/server/game/Battlefield/Battlefield.cpp]: maybe can rewrite it with deque and pop_front, what is more suitable?

**Comment by Noryad** [comment]
ready to merge and continue updating?

File: src/server/game/Battlefield/Battlefield.cpp (hunk 8 above)
Comment (Shauren) [reply src/server/game/Battlefield/Battlefield.cpp]: what is it currently?

File: src/server/game/Battlefield/Battlefield.h [hunk 9]
```diff
@@ -179,7 +180,7 @@ class TC_GAME_API Battlefield : public ZoneScript
         GuidUnorderedSet _players[PVP_TEAMS_COUNT];
         GuidUnorderedSet _playersInWar[PVP_TEAMS_COUNT];
         GuidUnorderedSet _groups[PVP_TEAMS_COUNT]; // contains the two different raid groups
-        GuidVector _playerQueue[PVP_TEAMS_COUNT];
+        std::deque<ObjectGuid> _playerQueue[PVP_TEAMS_COUNT];
```
Comment (Shauren) [inline src/server/game/Battlefield/Battlefield.h]: fyi - https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Entities/Object/ObjectGuid.h#L344
this also makes the include unneeded

Comment (ccrs) [reply src/server/game/Battlefield/Battlefield.h]: ow, didnt notice, ty

**Comment by nyhmii** [comment]
Latest Rev and commits:

Display issues : 

Possible 👍 

"
switch (_type)
	{
	case OBJECTTYPE_TOWER:
		UpdateCreatureAndGo();  <--- must there be a break; ? or not... since it updatesandgo's.. ?
	case OBJECTTYPE_KEEP_TOWER:
		UpdateTurretAttack(false);
		break;
	default:
		break;
	}
"

image

**Comment by ccrs** [comment]
I told you already I cannot reproduce what you are reporting, and that the description you are giving is really close to that of using a clean core.

**Comment by nyhmii** [comment]
@ccrs I've redownloaded a frehs core,compiled it. 2 hours ago. 

1. we tested the battle for about 5-6 times, worked perfectly. 7th time towers started changing faction colours and some walls didnt rebuild.

**Comment by ccrs** [comment]
https://ci.appveyor.com/project/DDuarte/trinitycore/build/1.0.6459/artifacts use that

**Comment by nyhmii** [comment]
@ccrs Thank you!

**Comment by ccrs** [comment]
thank @jackpoz its his idea  :)

**Comment by tje3d** [comment]
@ccrs i can't compile it :(
```
...
[100%] Linking CXX executable worldserver
../game/libgame.a(Opcodes.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrExitRequest::BattlefieldMgrExitRequest(WorldPacket&&)':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:89: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrExitRequest'
../game/libgame.a(Opcodes.cpp.o): In function `PacketHandler<WorldPackets::Battlefield::BattlefieldMgrExitRequest, &WorldSession::HandleBattlefieldExitRequest>::Call(WorldSession*, WorldPacket&) const':
/home/reza/TrinityCore/src/server/game/Server/Protocol/Opcodes.cpp:35: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrExitRequest::Read()'
../game/libgame.a(Opcodes.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrQueueInviteResponse::BattlefieldMgrQueueInviteResponse(WorldPacket&&)':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:64: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrQueueInviteResponse'
../game/libgame.a(Opcodes.cpp.o): In function `PacketHandler<WorldPackets::Battlefield::BattlefieldMgrQueueInviteResponse, &WorldSession::HandleBattlefieldQueueInviteResponse>::Call(WorldSession*, WorldPacket&) const':
/home/reza/TrinityCore/src/server/game/Server/Protocol/Opcodes.cpp:35: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrQueueInviteResponse::Read()'
../game/libgame.a(Opcodes.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrEntryInviteResponse::BattlefieldMgrEntryInviteResponse(WorldPacket&&)':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:42: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrEntryInviteResponse'
../game/libgame.a(Opcodes.cpp.o): In function `PacketHandler<WorldPackets::Battlefield::BattlefieldMgrEntryInviteResponse, &WorldSession::HandleBattlefieldEntryInviteResponse>::Call(WorldSession*, WorldPacket&) const':
/home/reza/TrinityCore/src/server/game/Server/Protocol/Opcodes.cpp:35: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrEntryInviteResponse::Read()'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrEntryInvite::BattlefieldMgrEntryInvite()':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:30: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrEntryInvite'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldSession::SendBattlefieldInvitePlayerToWar(unsigned int, unsigned int, unsigned int)':
/home/reza/TrinityCore/src/server/game/Handlers/BattlefieldHandler.cpp:42: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrEntryInvite::Write()'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrQueueInvite::BattlefieldMgrQueueInvite()':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:53: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrQueueInvite'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldSession::SendBattlefieldInvitePlayerToQueue(unsigned int)':
/home/reza/TrinityCore/src/server/game/Handlers/BattlefieldHandler.cpp:57: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrQueueInvite::Write()'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrQueueRequestResponse::BattlefieldMgrQueueRequestResponse()':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:75: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrQueueRequestResponse'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldSession::SendBattlefieldQueueInviteResponse(unsigned int, unsigned int, bool, bool)':
/home/reza/TrinityCore/src/server/game/Handlers/BattlefieldHandler.cpp:78: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrQueueRequestResponse::Write()'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrEntered::BattlefieldMgrEntered()':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:99: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrEntered'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldSession::SendBattlefieldEntered(unsigned int)':
/home/reza/TrinityCore/src/server/game/Handlers/BattlefieldHandler.cpp:95: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrEntered::Write()'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrEjected::BattlefieldMgrEjected()':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:112: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrEjected'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldSession::SendBattlefieldLeaveMessage(unsigned int, BFLeaveReason)':
/home/reza/TrinityCore/src/server/game/Handlers/BattlefieldHandler.cpp:113: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrEjected::Write()'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldPackets::Battlefield::BattlefieldMgrEjectPending::BattlefieldMgrEjectPending()':
/home/reza/TrinityCore/src/server/game/Server/Packets/BattlefieldPackets.h:125: undefined reference to `vtable for WorldPackets::Battlefield::BattlefieldMgrEjectPending'
../game/libgame.a(BattlefieldHandler.cpp.o): In function `WorldSession::SendBattlefieldEjectPending(unsigned int, bool)':
/home/reza/TrinityCore/src/server/game/Handlers/BattlefieldHandler.cpp:129: undefined reference to `WorldPackets::Battlefield::BattlefieldMgrEjectPending::Write()'
collect2: error: ld returned 1 exit status
src/server/worldserver/CMakeFiles/worldserver.dir/build.make:193: recipe for target 'src/server/worldserver/worldserver' failed
make[2]: *** [src/server/worldserver/worldserver] Error 1
CMakeFiles/Makefile2:1653: recipe for target 'src/server/worldserver/CMakeFiles/worldserver.dir/all' failed
make[1]: *** [src/server/worldserver/CMakeFiles/worldserver.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by tkrokli** [comment]
@tje3d : it could be an issue with the compiler version in your OS. Are you able to compile all other versions of TrinityCore source without errors or warnings? See also the output result from travis-ci : https://travis-ci.org/TrinityCore/TrinityCore/builds/273830099?utm_source=github_status&utm_medium=notification (you may want to expand the top information lines to see the various version info used by travis-ci). Please also compare with the support info from the wiki pages, in case you get issues when compiling other TC source as well.

**Comment by lachtanek** [comment]
@tje3d did you run cmake before compiling? Might cause something like this.

**Comment by tje3d** [comment]
@lachtanket ahh, its compiled succesfull. thanks <3

**Comment by lachtanek** [comment]
Found one new problem, tenacity on vehicles is only applied on player leave/enter.

**Comment by lachtanek** [comment]
Okay, I have some fixes:
- Tenacity should work correctly
- Teleporters now really teleport
- Fix worldwide WG aura
https://gist.github.com/lachtanek/bf5cfccd17f01dab3605c274022aad9c
https://gist.github.com/lachtanek/7770b94a2324a2c6fa6617cd14d1731c
We have some changes in our core so some line numbers may not be right on TC.

**Comment by tje3d** [comment]
@lachtanek here is patch's for clean repository:

- https://gist.github.com/tje3d/0587a8e3513583f21684681ace0cdb2e
- https://gist.github.com/tje3d/045c390885b098b48eb0fe36ed9e3d3d

**Comment by ccrs** [comment]
many thanks! I'll apply this with other changes as soon as I have some spare time

**Comment by Noryad** [comment]
We accidentally summoned a player who was in wintergrasp, apparently he kept a buff and it provoqued an exaggerated dps.

dps adidas maja

**Comment by lachtanek** [comment]
@ccrs any reason for this negation?

````void BattlefieldWintergrasp::OnPlayerLeaveZone(Player* player)
{
    if (!IsWarTime())
        RemoveAurasFromPlayer(player);
```

**Comment by GMKyle** [comment]
@ccrs Inside ICC, wintergrasp 's join notification is showing even if the player closed, this occurs until the battle ends. Also note in the previous image the remaining time and the Vehicles per faction. Zone, area or map problem?

**Comment by Dalgurakk** [comment]
Please can anyone confirm this?
1- Shortly after starting the event the tower on the right to the south breaks by itself.
2- When a player (engineer) with activated nitro boots uses a castle cannon and dismounts, he keeps the speed moving.

**Comment by sirikfoll** [comment]
@Dalgurakk  Bug number 2 is not related to wintergrasp, afaik it happens with any vehicle

**Comment by Noryad** [comment]
Wintergrasp is still disabled on my server, any (temp) fix for wintergrasp buff within ICC?

**Comment by hlainin** [comment]
Good evening. Can you please tell me which of these files implemented the script capturing Goblin workshops. Or it is implemented through the database? I apologize for this question in this discussion. I still poorly understand.

**Comment by Re3os** [comment]
@ccrs can not cancel invitation bar, click not work.

**Comment by Noryad** [comment]
how can i solve conflicts with 57672fd?

**Comment by tje3d** [comment]
@Noryad this will break Ruby Sanctum and Dreamwalker Phasing cause OnPlayerLeaveZone not working as expected ( just get called when you enter the zone xD )
Also cancel invitation not works

**Comment by Ascathor** [comment]
@tje3d It is preferable to add a diff on https://gist.github.com instead of a downloadable file for easier review.

**Comment by tje3d** [comment]
Could we open new PR to continue working on this?

**Comment by Keader** [comment]
@tje3d dont need, ccrs still working on it.

@ccrs  need fix Conflicts

**Comment by irgc** [comment]
hi
3.3.5 TC
wintergrasp fix?

**Comment by irgc** [comment]
i can enable wintergrasp? is  worked !

**Comment by ccrs** [comment]
no, probably around August I can continue this 👍

**Comment by irgc** [comment]
Since I remember it was a bug wintergrasp on trinitycore!
A bug was a long time ago
:smile:

**Comment by irgc** [comment]
WE WANT WiNTERGRASP :sob:

**Comment by ccrs** [comment]
AND I DONT HAVE ENOUGH FREE TIME

**Comment by ccrs** [comment]
I'm definately creating another PR for the continuation, the number of unrelated commnets is way too high

**Pull request closed without merging**